include/inexor/vulkan-renderer/world/collision.hpp

```diff
@@ -0,0 +1,250 @@
+            cube_edge{"middle bottom front", adjust_coordinates({0.0f, -1.0f, -1.0f})},
+            cube_edge{"middle bottom back", adjust_coordinates({0.0f, 1.0f, -1.0f})},
+            cube_edge{"middle top front", adjust_coordinates({0.0f, -1.0f, 1.0f})},
+            cube_edge{"middle top back", adjust_coordinates({0.0f, 1.0f, 1.0f})}};
+
+        using edge_on_face_index = std::array<std::size_t, 4>;
+
+        // These indices specify which 4 edges are associated with a given face.
+        static constexpr std::array<edge_on_face_index, 6> EDGE_ON_FACE_INDICES{
+            edge_on_face_index{0, 1, 2, 3},   // left
+            edge_on_face_index{4, 5, 6, 7},   // right
+            edge_on_face_index{1, 5, 8, 11},  // front
+            edge_on_face_index{3, 7, 9, 11},  // back
+            edge_on_face_index{0, 4, 10, 11}, // top
+            edge_on_face_index{2, 6, 8, 9}    // bottom
+        };
+
+        /// @brief Determine the intersection point between a ray and a plane.
+        /// @note This function is not available in glm.
+        /// @param plane_pos The position of the plane.
+        /// @param plane_norm The normal vector of the plane.
+        /// @param ray_pos Point a on the ray.
+        /// @param ray_dir Point b on the ray.
+        const auto ray_plane_intersection_point = [](const glm::vec3 plane_pos, const glm::vec3 plane_norm,
+                                                     const glm::vec3 ray_pos, const glm::vec3 ray_dir) {
+            return ray_pos - ray_dir * (glm::dot((ray_pos - plane_pos), plane_norm) / glm::dot(ray_dir, plane_norm));
+        };
+
+        /// @brief As soon as we determined which face is selected and we calculated the intersection point between the
+        /// ray and the face plane, we need to determine the nearest corner in that face and the nearest edge on that
+        /// face. In order to determine the nearest corner on a selected face for example, we would iterate through all
+        /// corners on the selected face and calculate the distance between the intersection point and the corner's
+        /// coordinates. The corner which is closest to the intersection point is the selected corner.
+        /// However, we should not use ``glm::distance`` for this, because it performs a ``sqrt`` calculation on the
+        /// vector coordinates. This is not necessary in this case, as we are not interested in the exact distance but
+        /// rather in a value which allows us to determine the nearest corner. This means we can use the squared
+        /// distance, which allows us to avoid the costly call of ``sqrt``.
+        /// @param pos1 The first point.
+        /// @param pos2 The second point.
+        const auto square_of_distance = [](const std::array<glm::vec3, 2> points) {
+            const auto diff = points[1] - points[0];
+            return static_cast<float>(std::pow(glm::dot(diff, diff), 2));
```

Comment:
Wouldn't the `std::pow` here make this function return `distance ^ 4`? I was going to suggest using `glm::distance2`, but they aren't doing the same thing in this case. I don't think this is really affecting anything though, since the distance is just used for comparison?

Reply:
Oh yes it does. The reason I didn't notice is because this also doesn't affect the relative order.

Reply:
There's a bug in this code. If I use `glm::distance`, the corner detection is correct. I need to check that squared distance calculation again.

```cpp
const auto square_of_distance = [&](const std::array<glm::vec3, 2> &points) {
    const auto diff = points[1] - points[0];
    return glm::distance(points[0],points[1]);
    //return static_cast<float>(std::pow(glm::dot(diff, diff), 2));
};
```

Reply:
Using `glm::distance2` fixed all issues. Done.